core/src/main/java/ai/timefold/solver/core/impl/domain/variable/provided/DefaultShadowCalculationBuilder.java

```diff
@@ -0,0 +1,114 @@
+package ai.timefold.solver.core.impl.domain.variable.provided;
+
+import java.util.function.BiFunction;
+
+import ai.timefold.solver.core.api.function.PentaFunction;
+import ai.timefold.solver.core.api.function.QuadFunction;
+import ai.timefold.solver.core.api.function.TriFunction;
+import ai.timefold.solver.core.impl.domain.solution.descriptor.SolutionDescriptor;
+import ai.timefold.solver.core.impl.domain.variable.supply.SupplyManager;
+import ai.timefold.solver.core.preview.api.variable.provided.ShadowCalculationBuilder;
+import ai.timefold.solver.core.preview.api.variable.provided.SingleVariableReference;
+import ai.timefold.solver.core.preview.api.variable.provided.VariableReference;
+
+public class DefaultShadowCalculationBuilder<Solution_, Entity_, Value_> implements ShadowCalculationBuilder<Entity_, Value_> {
```

Comment:
Also, you seem to be using the nullability markers somewhat selectively.
Why are they used in one place, and not in other places?

Reply:
Mostly, they are used in classes where nullability is mixed (i.e. some fields/methods can be null, others cannot).